[GStreamer][WebRTC] canvas to PeerConnection improvements

philn opened the pull request:
#### 88a6f746233889464d49f63561b84a00afac2411
<pre>
[GStreamer][WebRTC] canvas to PeerConnection improvements
<a href="https://bugs.webkit.org/show_bug.cgi?id=241330">https://bugs.webkit.org/show_bug.cgi?id=241330</a>

Patch by Philippe Normand &lt;philn@igalia.com &gt; on 2022-06-15
Reviewed by Miguel Gomez and Xabier Rodriguez-Calvar.

Video frames exported from the WebGL canvas need to be flipped on GStreamer ports and their format
should be RGBA. Frames coming from 2D canvas don&apos;t need to be flipped and have BGRA format.

Texture flipping is now handled in the player which is useful when painting canvas to a video
element. And flipping is also now handled in RTC outgoing video sources, which is useful when
exporting a canvas to a WebRTC PeerConnection.

The 30fps framerate is hardcoded, ideally should be inferred from the compositor refresh rate.
Having a framerate  &gt; 1 is specially needed for the canvas- &gt;WebRTC use-case, otherwise a 1/1
framerate would be set in caps, making the corresponding video stream slower than the actual refresh
rate.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::toVideoFrame):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateTextureMapperFlags):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrameGStreamer::createFromPixelBuffer):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapper.cpp:
(WebCore::GraphicsContextGLTextureMapper::paintCompositedResultsToVideoFrame):
* Source/WebCore/platform/mediastream/gstreamer/MockRealtimeVideoSourceGStreamer.cpp:
(WebCore::MockRealtimeVideoSourceGStreamer::updateSampleBuffer):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingVideoSourceGStreamer::RealtimeOutgoingVideoSourceGStreamer):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::setPayloadType):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/251560@main">https://commits.webkit.org/251560@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295555">https://svn.webkit.org/repository/webkit/trunk@295555</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

magomez commented on Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:
```diff
@@ -43,7 +49,7 @@ class VideoFrameGStreamer final : public VideoFrame {
     }
 
-    static Ref<VideoFrameGStreamer> createFromPixelBuffer(Ref<PixelBuffer>&&, const MediaTime& presentationTime = MediaTime::invalidTime(), const IntSize& destinationSize = { }, double frameRate = 1, Rotation videoRotation = Rotation::None, bool videoMirrored = false, std::optional<VideoFrameTimeMetadata>&& metadata = std::nullopt);
+    static Ref<VideoFrameGStreamer> createFromPixelBuffer(Ref<PixelBuffer>&&, const MediaTime& presentationTime = MediaTime::invalidTime(), const IntSize& destinationSize = { }, double frameRate = 1, Rotation videoRotation = Rotation::None, bool videoMirrored = false, std::optional<VideoFrameTimeMetadata>&& metadata = std::nullopt, CanvasContentType canvasContentType = CanvasContentType::WebGL);
```
It's weird that the default values for videoRotation and canvasContentType are incompatible. If canvasContentType is WebGL, then rotation should be UpsideDown, and if canvasContentType is Canvas2D then rotation should be none.

philn replied:
well there can be only one default value for each parameter, so not sure what to do here :)

magomez replied:
You can set Canvas2D as the default value for canvasContentType and that would match the default rotation of none. Or keep WebGL for canvasContentType and then set rotation to UpsideDown. The thing is that the default values as they are are not valid neither for Canvas2D not for WebGL.

philn replied:
maybe they should be mandatory args actually.

magomez replied:
that's also a good option.

magomez reviewed (commented):
Despite you did it properly on the implementation, the commit message is wrong. Frames Coming form Canvas2D are BGRA and should not be rotated, and frames coming from WebGL are RGBA and should be rotated.

aperezdc committed:
[WPE][GTK] REGRESSION(r292084): Browser frontend host is missing recent additions
https://bugs.webkit.org/show_bug.cgi?id=241594

Reviewed by Patrick Angle.

Update BrowserInspectorFrontendHost.js to add missing methods, update
signatures of changed ones, and remove those which are no longer present
in the IDL interface. This makes the Web Inspector work again when
loaded over HTTP.

* Source/WebInspectorUI/UserInterface/Base/BrowserInspectorFrontendHost.js:
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.canLoad): Added, always
returns false.
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.load): Added stub.
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.getPath): Added stub.
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.canPickColorFromScreen):
Added, always returns false.
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.pickColorFromScreen):
Added stub.
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.getCurrentX): Added stub.
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.getCurrentY): Added stub.
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.setPath): Added stub.
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.get port): Deleted.
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.append): Deleted.
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.close): Deleted.

Canonical link: https://commits.webkit.org/251559@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295554 268f45cc-cd09-0410-ab3c-d52691b4dbfc

philn committed:
[GStreamer][WebRTC] canvas to PeerConnection improvements
https://bugs.webkit.org/show_bug.cgi?id=241330

Patch by Philippe Normand <philn@igalia.com> on 2022-06-15
Reviewed by Miguel Gomez and Xabier Rodriguez-Calvar.

Video frames exported from the WebGL canvas need to be flipped on GStreamer ports and their format
should be RGBA. Frames coming from 2D canvas don't need to be flipped and have BGRA format.

Texture flipping is now handled in the player which is useful when painting canvas to a video
element. And flipping is also now handled in RTC outgoing video sources, which is useful when
exporting a canvas to a WebRTC PeerConnection.

The 30fps framerate is hardcoded, ideally should be inferred from the compositor refresh rate.
Having a framerate > 1 is specially needed for the canvas->WebRTC use-case, otherwise a 1/1
framerate would be set in caps, making the corresponding video stream slower than the actual refresh
rate.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::toVideoFrame):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateTextureMapperFlags):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrameGStreamer::createFromPixelBuffer):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapper.cpp:
(WebCore::GraphicsContextGLTextureMapper::paintCompositedResultsToVideoFrame):
* Source/WebCore/platform/mediastream/gstreamer/MockRealtimeVideoSourceGStreamer.cpp:
(WebCore::MockRealtimeVideoSourceGStreamer::updateSampleBuffer):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingVideoSourceGStreamer::RealtimeOutgoingVideoSourceGStreamer):
(WebCore::RealtimeOutgoingVideoSourceGStreamer::setPayloadType):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingVideoSourceGStreamer.h:

Canonical link: https://commits.webkit.org/251560@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295555 268f45cc-cd09-0410-ab3c-d52691b4dbfc
